src/lib.rs

```diff
@@ -217,22 +301,32 @@ where
     ///
     /// If the key already existed in the cache, the existing value is returned and overwritten in
     /// the cache.  Otherwise, the key-value pair is inserted and `None` is returned.
-    pub fn insert(&mut self, key: Key, value: Value) -> Option<Value> {
-        if self.map.contains_key(&key) {
+    /// Evicts and returns expired entries.
+    pub fn notify_insert(&mut self, key: Key, value: Value) -> (Option<Value>, Vec<(Key, Value)>) {
+        let expired = if self.map.contains_key(&key) {
             Self::update_key(&mut self.list, &key);
+            Vec::new()
         } else {
-            self.remove_expired();
-            if self.map.len() >= self.capacity {
-                for key in self.list.drain(..=self.map.len() - self.capacity) {
-                    assert!(self.map.remove(&key).is_some());
-                }
-            }
+            let expired = self.remove_expired();
```

Comment:
why do we only remove-expired in this case but return an empty vector in the case the key was found ? So if i insert into LRU something that existed, OK it's going to update it's timestamp, but why not also check and remove any expired stuff like in the case where i insert a non-existent key into the cache ?

Reply:
I tried not to change the existing code semantics. But that sounds good to me too.